FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1457 @@
+    configASSERT( xResult == MQTTSuccess );
+
+    /* Many fields not used in this demo so start with everything at 0. */
+    memset( ( void * ) &xConnectInfo, 0x00, sizeof( xConnectInfo ) );
+
+    /* Start with a clean session i.e. direct the MQTT broker to discard any
+     * previous session data. Also, establishing a connection with clean session
+     * will ensure that the broker does not store any data when this client
+     * gets disconnected. */
+    xConnectInfo.cleanSession = true;
+
+    /* The client identifier is used to uniquely identify this MQTT client to
+     * the MQTT broker. In a production device the identifier can be something
+     * unique, such as a device serial number. */
+    xConnectInfo.pClientIdentifier = CLIENT_IDENTIFIER;
+    xConnectInfo.clientIdentifierLength = ( uint16_t ) strlen( CLIENT_IDENTIFIER );
+
+    /* Set MQTT keep-alive period. It is the responsibility of the application to ensure
+     * that the interval between Control Packets being sent does not exceed the Keep Alive value.
+     * In the absence of sending any other Control Packets, the Client MUST send a PINGREQ Packet. */
+    xConnectInfo.keepAliveSeconds = MQTT_KEEP_ALIVE_INTERVAL_SECONDS;
+
+    /* Send MQTT CONNECT packet to broker. LWT is not used in this demo, so it
+     * is passed as NULL. */
+    xResult = MQTT_Connect( pxMQTTContext,
+                            &xConnectInfo,
+                            NULL,
+                            CONNACK_RECV_TIMEOUT_MS,
+                            &xSessionPresent );
+
+    if( xResult != MQTTSuccess )
+    {
+        LogError( ( "Connection with MQTT broker failed.\r\n" ) );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvInitializeCommandContext( CommandContext_t * pxContext )
+{
+    pxContext->complete = false;
+    pxContext->pResponseQueue = NULL;
+    pxContext->returnStatus = MQTTSuccess;
+    pxContext->pPublishInfo = NULL;
+    pxContext->pSubscribeInfo = NULL;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvDestroyCommandContext( CommandContext_t * pxContext )
+{
+    ( void ) pxContext;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvAddAck( uint16_t usPacketId,
+                       CommandContext_t * pxContext,
+                       CommandCallback_t xCallback )
+{
+    int32_t i = 0;
+
+    for( i = 0; i < PENDING_ACKS_MAX_SIZE; i++ )
+    {
+        if( pxPendingAcks[ i ].packetId == MQTT_PACKET_ID_INVALID )
+        {
+            pxPendingAcks[ i ].packetId = usPacketId;
+            pxPendingAcks[ i ].pCommandContext = pxContext;
+            pxPendingAcks[ i ].callback = xCallback;
+            break;
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static AckInfo_t prvPopAck( uint16_t usPacketId )
```

Comment:
I kind of assumed this is a stack the operation is named pop. Would prvRemoveAck be a better name for this function?

Reply:
Yes, it would, changed